src/RewardsStreamerMP.sol

```diff
@@ -41,15 +45,88 @@ contract RewardsStreamerMP is IStakeManager, TrustedCodehashAccess, ReentrancyGu
+    mapping(address owner => address[] vault) public vaults;
+    mapping(address vault => address owner) public vaultOwners;
+    mapping(address vault => Account data) public accounts;
+
+    modifier onlyRegisteredVault() {
+        if (!isVaultRegistered(msg.sender)) {
+            revert StakingManager__VaultNotRegistered();
+        }
+        _;
+    }
 
     constructor(address _owner, address _stakingToken, address _rewardToken) TrustedCodehashAccess(_owner) {
         STAKING_TOKEN = IERC20(_stakingToken);
         REWARD_TOKEN = IERC20(_rewardToken);
         lastMPUpdatedTime = block.timestamp;
     }
 
-    function stake(uint256 amount, uint256 lockPeriod) external onlyTrustedCodehash nonReentrant {
+    /**
+     * @notice Check if a vault is registered
+     * @param vault The address of the vault to check
+     * @return true if the vault is registered, false otherwise
+     */
+    function isVaultRegistered(address vault) public view returns (bool) {
+        return vaultOwners[vault] != address(0);
+    }
+
+    /**
+     * @notice Registers a vault with its owner. Called by the vault itself during initialization.
+     * @dev Only callable by contracts with trusted codehash
+     */
+    function registerVault() external onlyTrustedCodehash {
+        address vault = msg.sender;
+        address owner = IStakeVault(vault).owner();
+
+        if (vaultOwners[vault] != address(0)) {
+            revert StakingManager__VaultAlreadyRegistered();
+        }
+
+        // Verify this is a legitimate vault by checking it points to us
+        if (address(IStakeVault(vault).stakeManager()) != address(this)) {
+            revert StakingManager__InvalidVault();
+        }
+
+        vaultOwners[vault] = owner;
+        vaults[owner].push(vault);
+    }
+
+    function getUserVaults(address owner) external view returns (address[] memory) {
+        return vaults[owner];
+    }
+
+    function getTotalMP(address owner) external view returns (uint256) {
+        address[] memory userVaults = vaults[owner];
+        uint256 userTotalMP = 0;
+
+        for (uint256 i = 0; i < userVaults.length; i++) {
+            userTotalMP += accounts[userVaults[i]].accountMP + _accruedMP(userVaults[i]);
+        }
+        return userTotalMP;
+    }
+
+    function getTotalMaxMP(address owner) external view returns (uint256) {
+        address[] memory userVaults = vaults[owner];
+        uint256 userTotalMaxMP = 0;
+
+        for (uint256 i = 0; i < userVaults.length; i++) {
+            userTotalMaxMP += accounts[userVaults[i]].maxMP;
+        }
+        return userTotalMaxMP;
+    }
+
+    function getTotalStakedBalance(address owner) external view returns (uint256) {
+        address[] memory userVaults = vaults[owner];
+        uint256 userTotalStake = 0;
+
+        for (uint256 i = 0; i < userVaults.length; i++) {
+            userTotalStake += accounts[userVaults[i]].stakedBalance;
+        }
+        return userTotalStake;
+    }
```

Comment:
Any strong opinions on how to name these? Tried to stay consistent.

Also need to add natspec

Reply:
I would use something like `getUserXYZ` just to specify it's for a user instead of having something that sounds global like `getTotalMaxMP`